packages/grafana-ui/src/components/Graph/GraphWithLegend.tsx

```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+
+import { css } from 'emotion';
+import { Graph, GraphProps } from './Graph';
+import { LegendRenderOptions, LegendItem } from '../Legend/Legend';
+import { GraphLegend } from './GraphLegend';
+import { CustomScrollbar } from '../CustomScrollbar/CustomScrollbar';
+import { GraphSeriesValue } from '../../types/graph';
+
+export type SeriesOptionChangeHandler<TOption> = (label: string, option: TOption) => void;
+export type SeriesColorChangeHandler = SeriesOptionChangeHandler<string>;
+export type SeriesAxisToggleHandler = SeriesOptionChangeHandler<boolean>;
+
+export interface GraphWithLegendProps extends GraphProps, LegendRenderOptions {
+  decimals?: number;
+  isLegendVisible: boolean;
+  renderLegendAsTable: boolean;
+  sortLegendBy?: string;
+  sortLegendDesc?: boolean;
+  onSeriesColorChange: SeriesColorChangeHandler;
+  onSeriesAxisToggle?: SeriesAxisToggleHandler;
+  onSeriesToggle?: (label: string, event: React.MouseEvent<HTMLElement>) => void;
+  onToggleSort: (sortBy: string) => void;
+}
+
+const getGraphWithLegendStyles = ({ placement }: GraphWithLegendProps) => ({
+  wrapper: css`
+    display: flex;
+    flex-direction: ${placement === 'under' ? 'column' : 'row'};
+    height: 100%;
+  `,
+  graphContainer: css`
+    min-height: 65%;
+    flex-grow: 1;
+  `,
+  legendContainer: css`
+    padding: 10px 0;
+    max-height: ${placement === 'under' ? '35%' : 'none'};
+  `,
+});
+
+const shouldHideLegendItem = (data: GraphSeriesValue[][], hideEmpty = false, hideZero = false) => {
+  const isZeroOnlySeries = data.reduce((acc, current) => acc + (current[1] || 0), 0) === 0;
+  const isNullOnlySeries = !data.reduce((acc, current) => acc && current[1] !== null, true);
+
+  return (hideEmpty && isNullOnlySeries) || (hideZero && isZeroOnlySeries);
+};
+
+export const GraphWithLegend: React.FunctionComponent<GraphWithLegendProps> = (props: GraphWithLegendProps) => {
+  const {
+    series,
+    timeRange,
+    width,
+    height,
+    showBars,
+    showLines,
+    showPoints,
+    sortLegendBy,
+    sortLegendDesc,
+    isLegendVisible,
+    renderLegendAsTable,
+    placement,
+    onSeriesAxisToggle,
+    onSeriesColorChange,
+    onSeriesToggle,
+    onToggleSort,
+    hideEmpty,
+    hideZero,
+  } = props;
+  const { graphContainer, wrapper, legendContainer } = getGraphWithLegendStyles(props);
+
+  return (
+    <div className={wrapper}>
+      <div className={graphContainer}>
+        <Graph
+          series={series.filter(s => !!s.isVisible)}
+          timeRange={timeRange}
+          showLines={showLines}
+          showPoints={showPoints}
+          showBars={showBars}
+          width={width}
+          height={height}
+          key={isLegendVisible ? 'legend-visible' : 'legend-invisible'}
+        />
+      </div>
+
+      {isLegendVisible && (
+        <div className={legendContainer}>
+          <CustomScrollbar renderTrackHorizontal={props => <div {...props} style={{ visibility: 'none' }} />}>
```

Comment:
wonder if this can be done some other way, feels hacky

Reply:
note to myself: use bool prop

Reply:
note to myself: use bool prop